WalletWasabi/Affiliation/AffiliateInformationUpdater.cs

```diff
@@ -0,0 +1,297 @@
+using System.Linq;
+using System.Collections.Immutable;
+using System.Collections.Generic;
+using WalletWasabi.Affiliation.Models;
+using NBitcoin;
+using WalletWasabi.WabiSabi.Backend.Rounds;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.WabiSabi.Backend.Events;
+using WalletWasabi.Bases;
+using System.Collections.Concurrent;
+using WalletWasabi.Affiliation.Models.CoinjoinRequest;
+
+namespace WalletWasabi.Affiliation;
+
+public class CoinjoinRequestsUpdater : PeriodicRunner
+{
+	private static TimeSpan AffiliateServerTimeout = TimeSpan.FromSeconds(60);
+	private static TimeSpan Interval = TimeSpan.FromSeconds(1);
+
+	public CoinjoinRequestsUpdater(Arena arena, ImmutableDictionary<AffiliationFlag, AffiliateServerHttpApiClient> clients, Signer signer)
+		  : base(Interval)
+	{
+		Arena = arena;
+		Clients = clients;
+		Signer = signer;
+
+		CoinjoinRequests = new();
+		RoundData = new();
+		RoundsToUpdate = new();
+		RoundsToRemove = new();
+
+		AddHandlers(arena);
+	}
+
+	private Arena Arena { get; }
+	private Signer Signer { get; }
+	private ImmutableDictionary<AffiliationFlag, AffiliateServerHttpApiClient> Clients { get; }
+	private Dictionary<uint256, Dictionary<AffiliationFlag, byte[]>> CoinjoinRequests { get; }
+	private ConcurrentDictionary<uint256, RoundData> RoundData { get; }
+	private Queue<uint256> RoundsToUpdate { get; }
+	private Queue<uint256> RoundsToRemove { get; }
+
+	public override void Dispose()
+	{
+		RemoveHandlers(Arena);
+	}
+
+	public IReadOnlyDictionary<uint256, IReadOnlyDictionary<AffiliationFlag, byte[]>> GetCoinjoinRequests()
+	{
+		return (IReadOnlyDictionary<uint256, IReadOnlyDictionary<AffiliationFlag, byte[]>>)CoinjoinRequests.ToDictionary(x => x.Key, x => (IReadOnlyDictionary<AffiliationFlag, byte[]>)x.Value);
+	}
+
+	protected override async Task ActionAsync(CancellationToken cancel)
+	{
+		await UpdateCoinjoinRequestsAsync();
+		RemoveRounds();
+	}
+
+	private async Task UpdateCoinjoinRequestsAsync(uint256 roundId, RoundData roundData, AffiliationFlag affiliationFlag, AffiliateServerHttpApiClient affiliateServerHttpApiClient)
+	{
+		Body body = roundData.GetAffiliationData(affiliationFlag);
+		byte[]? result = await GetCoinjoinRequestAsync(affiliateServerHttpApiClient, body);
+
+		if (result is not null)
```

Comment:
Do not use null as a way to indicate that something was wrong because we lose the information about what was wrong. The the exception to bubble and catch it here were we can log the problem.

Reply:
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/0b6565315458c4fc1e1b4335ac2ef357cddb7b38.